[New Term Entry] Java Math Methods: `.ulp()` (#1255)
Codecademy/docs: content/java/concepts/math-methods/terms/ulp/ulp.md

---
Title: '.ulp()'
Description: 'Returns the size of the unit of last precision of a number.'
Subjects:
  - 'Computer Science'
Tags:
  - 'Methods'
  - 'Arithmetic'
  - 'Documentation'
  - 'Numbers'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

> The **Math.ulp()** method returns the unit of least precision of a given number. It calculates the distance between the given `float` or `double` value and the next largest `float` or `double` value in magnitude.

Reviewer: ```suggestion
The **`Math.ulp()`** method returns the unit of least precision of a given number. It calculates the distance between the given `float` or `double` value and the next largest `float` or `double` value in magnitude.
```

## Syntax

```pseudo
Math.ulp(num);
```

The `num` parameter can be of type `float` or `double` and the return value will be of the same type.

Special cases include the following:

- If `num` is `NaN`, `NaN` will be returned.
- If `num` is positive or negative zero, then the return value will be `Double.MIN_VALUE` or `Float.MIN_VALUE` which is the lowest number a `float` or `double` can represent.
- If `num` is positive or negative infinity, then the return value will be postive infinity.

> - If `num` is a number, then `num` will have the same return value as `-num`

Reviewer: ```suggestion
- If `num` is a number, then `num` will have the same return value as `-num`.
```

- If `num` is `Double.MAX_VALUE` or `Float.MAX_VALUE`, then the return value will be 2<sup>971</sup> for `double` and 2<sup>104</sup> for `float`.

## Example 1

> ```java

Reviewer: ```suggestion
```java
// Test.java
```

> import java.lang.Math

Reviewer: ```suggestion
import java.lang.Math;
```

public class Test {
  public static void main(String args[]) {

>     float num = 23.44;

Reviewer: ```suggestion
    double num = 23.44;
```

    System.out.println(Math.ulp(num));
  }
}
```

This will output:

```shell
3.552713678800501E-15
```

## Example 2

```java

> import java.lang.Math

Reviewer: ```suggestion
import java.lang.Math;
```

public class Test {
  public static void main(String args[]) {

>     float num = 23.44;

Reviewer: ```suggestion
    double num = 23.44;
```

---
Title: '.ulp()'
Description: 'Returns the size of the unit of last precision of a number.'
Subjects:
  - 'Computer Science'
Tags:
  - 'Methods'
  - 'Arithmetic'
  - 'Documentation'
  - 'Numbers'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

The **Math.ulp()** method returns the unit of least precision of a given number. It calculates the distance between the given `float` or `double` value and the next largest `float` or `double` value in magnitude.

## Syntax

```pseudo
Math.ulp(num);
```

> The `num` parameter can be of type `float` or `double`.
> 
> - The return type will be the same as the parameter type.

Reviewer: ```suggestion
The `num` parameter can be of type `float` or `double` and the return type will be of the same type.
```

Special cases include the following:

- If `num` is `NaN`, `NaN` will be returned.
- If `num` is positive or negative zero, then the return value will be `Double.MIN_VALUE` or `Float.MIN_VALUE` which is the lowest number a `float` or `double` can represent.
- If `num` is positive or negative infinity, then the return value will be postive infinity.
- If `num` is a number, then `num` will have the same return value as `-num`

> - If the argument is `Double.MAX_VALUE` or `Float.MAX_VALUE`, then the return value will be 2<sup>971</sup> for double and 2<sup>104</sup> for float.

Reviewer: ```suggestion
- If `num` is `Double.MAX_VALUE` or `Float.MAX_VALUE`, then the return value will be 2<sup>971</sup> for `double` and 2<sup>104</sup> for `float`.
```

---
Title: '.ulp()'

> Description: 'Returns the size of the ulp of a number which is its unit of last precision.'

Reviewer: ```suggestion
Description: 'Returns the size of the unit of last precision of a number.'
```

Subjects:
  - 'Computer Science'
Tags:
  - 'Methods'
  - 'Arithmetic'
  - 'Documentation'
  - 'Numbers'
CatalogContent:
  - 'learn-java'
  - 'paths/computer-science'
---

> The **Math.ulp()** method returns the unit of least precision of a given number. It calculates the distance between the given `float` or `double` value and the `float` `double` value next larger in magnitude.

Reviewer 2: ```suggestion
The **Math.ulp()** method returns the unit of least precision of a given number. It calculates the distance between the given `float` or `double` value and the next largest `float` or `double` value in magnitude.
```

## Syntax

```pseudo
Math.ulp(num);
```

> - The `num` parameter can either be of type `float` or `double`.

Reviewer 2: ```suggestion
The `num` parameter can be of type `float` or `double`.
```

- The return type will be the same as the parameter type.

Special cases include the following:

> - If `num` is NaN `ulp(num)` will be NaN.

Reviewer: ```suggestion
- If `num` is `NaN`, `NaN` will be returned.
```

> - If `num` is positive or negative zero, then the return value will be `Double.MAX_VALUE` or `Float.MAX_VALUE` which is the minimum numbers that `float` or `double` can represent.

Reviewer 2: The return value for a positive or negative zero is *`Double.MIN_VALUE`* (or *`Float.MIN_VALUE`*), the lowest number a `float` or `double` can represent.

https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#ulp-double-

Author: Of course I will fix that asap

Author: done now it should be correct! also I used a slightly different phrasing.

> - If `num` is positive or negative Zero then `ulp(num)` will be `Double.MAX_VALUE` or `Float.MAX_VALUE` meaning the minimum numbers that `float` or `double` can represent.

Reviewer: ```suggestion
- If `num` is positive or negative zero, then the return value will be `Double.MAX_VALUE` or `Float.MAX_VALUE` which is the minimum numbers that `float` or `double` can represent.
```

- If `num` is positive or negative infinity `ulp(num)` will be postive infinity.
- If `num` is a number `ulp(num)` is equal to `ulp(-num)`.

> - If the argument is `Double.MAX_VALUE` or `Float.MAX_VALUE` `ulp(num)` will be 2<sup>971</sup> for double and 2<sup>104</sup> for float.

Reviewer: Rather than stating the method itself each time, let's refer to it as the return value

Author: I agree it's more clear and readable

## Example 1

```java

> // ulp of a number

Reviewer: Include the file name as a comment `// Test.java`

import java.lang.Math

public class Test {
  public static void main(String args[]) {
    float num = 23.44;
    System.out.println(Math.ulp(num));
  }
}
```

> Output:

Reviewer: ```suggestion
This will output:
```

```shell
3.552713678800501E-15
```

## Example 2

```java
// some special cases
import java.lang.Math

public class Test {
  public static void main(String args[]) {
    float num = 23.44;
    System.out.println(Math.ulp(num) + " = "  + Math.ulp(-num));
    System.out.println(Math.ulp(0/0.));
    System.out.println(Math.ulp(0.));
    System.out.println(Math.ulp(1./0));
  }
}
```

> Output:

Reviewer: ```suggestion
This will output:
```
